extra_data/calibration.py

```diff
@@ -0,0 +1,1090 @@
+            metadata (CCVMetadata, optional): CCV metadata to load
+                constant data for, may be None to query metadata.
+
+        Returns:
+            (ndarray): CCV data
+        """
+
+        if self.caldb_root is None:
+            raise RuntimeError('calibration database store unavailable')
+
+        if self.modules and module not in self.modules:
+            raise ValueError('module not part of this correction data')
+
+        if metadata is None:
+            metadata = self.metadata([calibration])
+
+        row = metadata[module][calibration]
+
+        with h5py.File(self.caldb_root / row['path'], 'r') as f:
+            return np.asarray(f[row['dataset'] + '/data'])
+
+    def ndarray_map(self, calibrations=None, metadata=None):
+        """Load all CCV data in a nested map of ndarrays.
+
+        Args:
+            calibrations (Iterable of str, optional): Calibration constants
+                or None for all available (default).
+            metadata (CCVMetadata, optional): CCV metadata to load constant
+                for or None to query metadata automatically (default).
+
+        Returns:
+            (dict of dict of ndarray): CCV data by module number and
+                calibration constant name.
+        """
+
+        if self.caldb_root is None:
+            raise RuntimeError('calibration database store unavailable')
+
+        if metadata is None:
+            metadata = self.metadata(calibrations)
+
+        return {modno: {name: self.ndarray(modno, name, metadata)
+                        for name in data.keys()}
+                for modno, data in metadata.items()}
+
+    def _build_condition(self, parameters):
+        cond = dict()
+
+        for db_name in parameters:
+            value = getattr(self, self._simplify_parameter_name(db_name), None)
+
+            if value is not None:
+                cond[db_name] = value
+
+        return cond
+
+    @classmethod
+    def _from_multimod_detector_data(cls, component_cls, data, detector,
+                                     modules, client):
+        if isinstance(detector, component_cls):
+            detector_name = detector.detector_name
+        elif detector is None:
+            detector_name = component_cls._find_detector_name(data)
+        elif isinstance(detector, str):
+            detector_name = detector
+        else:
+            raise ValueError(f'detector may be an object of type '
+                             f'{type(cls)}, a string or None')
+
+        source_to_modno = dict(component_cls._source_matches(
+            data, detector_name))
+        detector_sources = [data[source] for source in source_to_modno.keys()]
+
+        if modules is None:
+            modules = sorted(source_to_modno.values())
+
+        creation_date = cls._determine_data_creation_date(data)
+
+        # Create new CorrectionData object.
+        corrdata = cls(detector_name, modules, client,
+                       creation_date, creation_date)
```

Comment:
This is using `creation_date` for both `event_at` and `snapshot_at`, which IIUC means it will only consider CCVs injected before the run was taken. I'd expect `snapshot_at=None` to be the default, so it would also consider CCVs injected afterwards.

Tracing through the code paths used to retrieve constants for the offline correction at the moment, I *think* it's setting `snapshot_at` for identifying the PDUs, and then unsetting it again before actually looking up the CCV. Specifically, here:

https://git.xfel.eu/karaboDevices/calibrationDBRemote/-/blob/3cce401db508d690e4998eab87a588b5bba52034/src/calibrationDBRemote/calibration_db_remote.py#L395

(And figuring that out reminds me that we should really move this forwards and get rid of calibrationDBRemote and iCalibrationDB ASAP)